### PR TITLE
Provide list of products

### PR DESCRIPTION
**Pull request by maxarvid**
PT Feature link: https://www.pivotaltracker.com/story/show/179716946

- Product model with attributes name, price, description of ingredients and image
- Create a show method
- GET request
```
As a restaurant API
In order to serve a client with products
I would like to be able to provide an endpoint for a products index

### CODE REVIEW

File: app/controllers/api/products_controller.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class Api::ProductsController < ApplicationController
+  def index
+    products = Product.all
+    render json: { products: products }
+  end
+
+  def show
+    product = Product.find(params['id'])
+    render json: { product: product }
+  end
```
Comment (offcast-dev) [inline app/controllers/api/products_controller.rb]: So judging by user story and PR title you need to provide a list of products in this feature. Show method returns individual product so it is out of scope on this one, you should have a separate PR for that. The smaller the PRs the better.

File: config/routes.rb [hunk 2]
```diff
@@ -1,7 +1,11 @@
 Rails.application.routes.draw do
   namespace :api do
+    get 'products/index'
+  end
```
Comment (offcast-dev) [inline config/routes.rb]: What does that do? not sure if you need it

File: config/routes.rb [hunk 3]
```diff
@@ -1,7 +1,11 @@
...
+  namespace :api do
+    resources :products, only: %i[index show]
     namespace :v0 do
       resources :pings, only: [:index], constraints: { format: 'json' }
```
Comment (offcast-dev) [inline config/routes.rb]: same with this, what are pings?

File: spec/requests/api/api_responds_with_a_specific_article_spec.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    it 'is expected to return the description of the requested product' do
+      expect(response_json['product']['description']).to eq "Delicious whale fat"
+    end
+
+    it 'is expected to return an image url for the requested product' do
+      expect(response_json['product']['image']).to eq "stringURL"
+    end
+  end
+end
```
Comment (offcast-dev) [inline spec/requests/api/api_responds_with_a_specific_article_spec.rb]: Why spec is called specific_article?

File: db/migrate/20210928144651_add_image_to_product.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddImageToProduct < ActiveRecord::Migration[6.0]
+  def change
+    add_column :products, :image, :string
+  end
+end
```
Comment (offcast-dev) [inline db/migrate/20210928144651_add_image_to_product.rb:5]: Look up how to create migrations with all attributes (price, name, ...) in one migration rater than having separate migration per attribute. It is ok now but will be pain later on

**Review by offcast-dev — CHANGES_REQUESTED** [review]
Change spec title and answer some questions. PR also has scope problems. And a last thing, include user story in the PR description.

File: app/controllers/api/products_controller.rb (hunk 1 above)
Comment (johnnysee) [reply app/controllers/api/products_controller.rb]: Noted, we rewrote show to return all products.

File: config/routes.rb (hunk 2 above)
Comment (johnnysee) [reply config/routes.rb]: It was auto generated by Rails, but we have removed it now.

File: config/routes.rb (hunk 3 above)
Comment (johnnysee) [reply config/routes.rb]: It was auto generated by Rails, but we have removed it now.

File: spec/requests/api/api_responds_with_a_specific_article_spec.rb (hunk 4 above)
Comment (johnnysee) [reply spec/requests/api/api_responds_with_a_specific_article_spec.rb]: This spec was out of scope, we are removing it.

File: db/migrate/20210928144651_add_image_to_product.rb (hunk 5 above)
Comment (johnnysee) [reply db/migrate/20210928144651_add_image_to_product.rb:5]: Noted, we will issue this next time we do it.

We will do this next time:

`rails generate migration AddDetailsToProjects price:decimal description:text image:string`

File: spec/requests/api/api_responds_with_a_specific_article_spec.rb (hunk 4 above)
Comment (maxarvid) [reply spec/requests/api/api_responds_with_a_specific_article_spec.rb]: Changing name to specific_product.

File: app/controllers/api/products_controller.rb (hunk 1 above)
Comment (maxarvid) [reply app/controllers/api/products_controller.rb]: Thomas asks to keep index method in and keep its functionality of returning an index of products.

**Review by offcast-dev — APPROVED** [review]
Don't forget about scope of the PR

**Review by tochman — COMMENTED** [review]
You are still not returning the same object as the front-end team is using in their PR. Is there another story in your PT that deals with the categories? Will that be updated later?

**Comment by maxarvid** [comment]
Yes, here is the link: https://www.pivotaltracker.com/story/show/179716949

**Pull request merged**